pkg/k8s/policyhelper/policy.go

```diff
@@ -119,3 +156,290 @@ func conflictResolutionSort[T core.Policy](policies []T) {
+	return &k8sPolicyClient[T, U, P, PL]{client: c}
+}
+
+func (pc *k8sPolicyClient[T, U, P, PL]) newList() PL {
+	var u U
+	return &u
+}
+
+func (pc *k8sPolicyClient[T, U, P, PL]) newPolicy() P {
+	var t T
+	return &t
+}
+
+func (pc *k8sPolicyClient[T, U, P, PL]) List(ctx context.Context, namespace string) ([]P, error) {
+	l := pc.newList()
+	opts := &k8sclient.ListOptions{}
+	if namespace != "" { // policies suppose to be namespaced
+		opts.Namespace = namespace
+	}
+	err := pc.client.List(ctx, l, opts)
+	if err != nil {
+		return nil, err
+	}
+	return l.GetItems(), nil
+}
+
+func (pc *k8sPolicyClient[T, U, P, PL]) Get(ctx context.Context, nsname types.NamespacedName) (P, error) {
+	p := pc.newPolicy()
+	err := pc.client.Get(ctx, nsname, p)
+	return p, err
+}
+
+func (pc *k8sPolicyClient[T, U, P, PL]) TargetRefObj(ctx context.Context, p P) (k8sclient.Object, error) {
+	tr := p.GetTargetRef()
+	obj := GroupKindObj(TargetRefGroupKind(tr))
+	key := types.NamespacedName{
+		Namespace: p.GetNamespace(),
+		Name:      string(tr.Name),
+	}
+	err := pc.client.Get(ctx, key, obj)
+	if err != nil {
+		return nil, err
+	}
+	return obj, nil
+}
+
+func (pc *k8sPolicyClient[T, U, P, PL]) UpdateStatus(ctx context.Context, policy P) error {
+	return pc.client.Status().Update(ctx, policy)
+}
+
+// Get all policies for given object, filtered by targetRef match.
+// List might include invalid and conflicting policies, for non-conflicting policy use ObjResolvedPolicy
+func (h *PolicyHandler[P]) ObjPolicies(ctx context.Context, obj k8sclient.Object) ([]P, error) {
+	allPolicies, err := h.client.List(ctx, obj.GetNamespace())
+	if err != nil {
+		return nil, err
+	}
+	out := []P{}
+	for _, policy := range allPolicies {
+		tr := policy.GetTargetRef()
+		if h.targetRefMatch(obj, tr) {
+			out = append(out, policy)
+		}
+	}
+	return out, nil
+}
+
+// Get resolved policy. Returns policy with conflict resolution. Will return at most single policy.
+// Every policy is implicitly Direct Policy Attachment and having more that 1 will result in conflict.
+func (h *PolicyHandler[P]) ObjResolvedPolicy(ctx context.Context, obj k8sclient.Object) (P, error) {
+	var empty P
+	objPolicies, err := h.ObjPolicies(ctx, obj)
+	if err != nil {
+		return empty, err
+	}
+	if len(objPolicies) == 0 {
+		return empty, nil
+	}
+	h.conflictResolutionSort(objPolicies)
+	return objPolicies[0], nil
+}
+
+// Add Watchers for configured Kinds to controller builder
+func (h *PolicyHandler[P]) AddWatchers(b *builder.Builder, objs ...k8sclient.Object) {
+	h.log.Debugf("add watchers for types: %v", NewGroupKindSet(objs).Items())
+	for _, watchObj := range objs {
+		b.Watches(watchObj, handler.EnqueueRequestsFromMapFunc(h.watchMapFn))
+	}
+}
+
+func (h *PolicyHandler[P]) watchMapFn(ctx context.Context, obj k8sclient.Object) []reconcile.Request {
+	out := []reconcile.Request{}
+	policies, err := h.client.List(ctx, obj.GetNamespace())
```

Comment:
Change name from `h.client` to `h.policyClient` is actually more clear for me

Reply:
There is single client instance in policy and it's PolicyClient. There is no ambiguity in here